src/goto-instrument/contracts/dynamic-frames/dfcc_contract_handler.cpp

```diff
@@ -110,39 +110,58 @@ void dfcc_contract_handlert::add_contract_instructions(
     .add_to_dest(dest, function_pointer_contracts);
 }
 
-const symbolt &
-dfcc_contract_handlert::get_pure_contract_symbol(const irep_idt &contract_id)
+const symbolt &dfcc_contract_handlert::get_pure_contract_symbol(
+  const irep_idt &contract_id,
+  const optionalt<irep_idt> function_id_opt)
 {
-  const auto &contract_symbol = utils.get_function_symbol(contract_id);
   auto pure_contract_id = "contract::" + id2string(contract_id);
   const symbolt *pure_contract_symbol = nullptr;
   if(!ns.lookup(pure_contract_id, pure_contract_symbol))
   {
-    check_signature_compat(
-      contract_id,
-      to_code_type(contract_symbol.type),
-      pure_contract_id,
-      to_code_type(pure_contract_symbol->type));
+    if(function_id_opt.has_value())
+    {
+      auto function_id = function_id_opt.value();
+      const auto &function_symbol = utils.get_function_symbol(function_id);
+      check_signature_compat(
+        function_id,
+        to_code_type(function_symbol.type),
+        pure_contract_id,
+        to_code_type(pure_contract_symbol->type));
+    }
     return *pure_contract_symbol;
   }
   else
   {
     // The contract symbol might not have been created if the function had
     // no contract or a contract with all empty clauses (which is equivalent).
-    // in that case we create a fresh symbol again with empty clauses
+    // in that case we create a fresh symbol again with empty clauses.
+    PRECONDITION_WITH_DIAGNOSTICS(
+      function_id_opt.has_value(),
+      "Contract " + pure_contract_id +
+        "not found, the identifier of an existing function must be provided to "
+        "derive a default contract");
+
+    auto function_id = function_id_opt.value();
+    const auto &function_symbol = utils.get_function_symbol(function_id);
+
+    log.warning() << "Contract '" << contract_id
+                  << "' not found, deriving empty pure contract '"
+                  << pure_contract_id << "' from function '" << function_id
+                  << "'" << messaget::eom;
+
     symbolt new_symbol;
     new_symbol.name = pure_contract_id;
     new_symbol.base_name = pure_contract_id;
     new_symbol.pretty_name = pure_contract_id;
     new_symbol.is_property = true;
-    new_symbol.type = contract_symbol.type;
-    new_symbol.mode = contract_symbol.mode;
-    new_symbol.module = contract_symbol.module;
-    new_symbol.location = contract_symbol.location;
+    new_symbol.type = function_symbol.type;
+    new_symbol.mode = function_symbol.mode;
+    new_symbol.module = function_symbol.module;
+    new_symbol.location = function_symbol.location;
```

Comment:
Can you also change this code to make use of a non-trivial `symbolt` constructor that's now available?

Reply:
now using symbolt{name, type, mode}